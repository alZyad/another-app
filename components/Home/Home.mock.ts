export const defaultSearch = {
  error: false,
  amount: 10,
  jokes: [
    {
      category: "Programming",
      type: "twopart",
      setup: "How many programmers does it take to screw in a light bulb?",
      delivery: "None. It's a hardware problem.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 1,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Have a great weekend!\nI hope your code behaves the same on Monday as it did on Friday.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 44,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Debugging is like being the detective in a crime movie where you're also the murderer at the same time.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 42,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "Why are Assembly programmers always soaking wet?",
      delivery: "They work below C-level.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      safe: true,
      id: 264,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: 'Judge: "I sentence you to the maximum punishment..."\nMe (thinking): "Please be death, please be death..."\nJudge: "Learn Java!"\nMe: "Damn."',
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 45,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Your mama's so FAT she can't save files bigger than 4GB.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: true,
      },
      id: 9,
      safe: false,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "Why did the Python programmer not respond to the foreign mails he got?",
      delivery: "Because his interpreter was busy collecting garbage.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 15,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Documentation is like sex:\nWhen it's good, it's very good.\nWhen it's bad, it's better than nothing...",
      flags: {
        nsfw: true,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      safe: false,
      id: 305,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "What are bits?",
      delivery: "Tiny things left when you drop your computer down the stairs.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 211,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Being a self-taught developer is almost the same as being a cut neck chicken because you have no sense of direction in the beginning.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 184,
      safe: false,
      lang: "en",
    },
  ],
};

export const emptySearch = {
  error: true,
  internalError: false,
  code: 106,
  message: "No matching joke found",
  causedBy: ["No jokes were found that match your provided filter(s)."],
  additionalInfo: "Error while finalizing joke filtering: No jokes were found that match your provided filter(s).",
  timestamp: 1721145331890,
};

export const fullSearch = {
  error: false,
  amount: 10,
  jokes: [
    {
      category: "Programming",
      type: "single",
      joke: "The six stages of debugging:\n1. That can't happen.\n2. That doesn't happen on my machine.\n3. That shouldn't happen.\n4. Why does that happen?\n5. Oh, I see.\n6. How did that ever work?",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 123,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "How do you generate a random string?",
      delivery: "Put a Windows user in front of Vim and tell them to exit.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 127,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "Algorithm: A word used by programmers when they don't want to explain how their code works.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 51,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "Why was the JavaScript developer sad?",
      delivery: "Because they didn't Node how to Express themself!",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      safe: true,
      id: 292,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: "How do you tell HTML from HTML5?\n- Try it out in Internet Explorer\n- Did it work?\n- No?\n- It's HTML5.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 43,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "How many programmers does it take to screw in a light bulb?",
      delivery: "None. It's a hardware problem.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 1,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "How did the programmer die in the shower?",
      delivery: "He read the shampoo bottle instructions: Lather. Rinse. Repeat.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      safe: false,
      id: 266,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "How do you know God is a shitty programmer?",
      delivery: "He wrote the OS for an entire universe, but didn't leave a single useful comment.",
      flags: {
        nsfw: false,
        religious: true,
        political: false,
        racist: false,
        sexist: false,
        explicit: true,
      },
      id: 19,
      safe: false,
      lang: "en",
    },
    {
      category: "Programming",
      type: "single",
      joke: 'Four engineers get into a car. The car won\'t start.\nThe Mechanical engineer says "It\'s a broken starter".\nThe Electrical engineer says "Dead battery".\nThe Chemical engineer says "Impurities in the gasoline".\nThe IT engineer says "Hey guys, I have an idea: How about we all get out of the car and get back in".',
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 132,
      safe: true,
      lang: "en",
    },
    {
      category: "Programming",
      type: "twopart",
      setup: "How can you tell an extroverted programmer?",
      delivery: "He looks at YOUR shoes when he's talking.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      id: 17,
      safe: true,
      lang: "en",
    },
  ],
};

export const oneResultSearch = {
  error: false,
  amount: 1,
  jokes: [
    {
      category: "Programming",
      type: "single",
      joke: "I have a joke about Stack Overflow, but you would say it's a duplicate.",
      flags: {
        nsfw: false,
        religious: false,
        political: false,
        racist: false,
        sexist: false,
        explicit: false,
      },
      safe: true,
      id: 301,
      lang: "en",
    },
  ],
};
