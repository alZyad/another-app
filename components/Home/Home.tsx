import { useEffect, useState } from "react";
import { StyleSheet, View, TextInput } from "react-native";
import JokeList from "../JokeList/JokeList";
import EmptyList from "../EmptyList/EmptyList";
import { JokeType } from "../../types/joke.types";

const getJokes = async (searchTerm: string) => {
  try {
    const response = await fetch(`https://v2.jokeapi.dev/joke/Programming?contains=${searchTerm}&amount=10`);
    const json = await response.json();
    return json;
  } catch (error) {
    console.error(">", error);
  }
};

export default function Home() {
  const [searchTerm, setSearchTerm] = useState("");
  const [data, setData] = useState<{ error: boolean; jokes: JokeType[]; message: string }>();

  useEffect(() => {
    getJokes(searchTerm).then((receivedData) => setData(receivedData));
  }, [searchTerm]);

  if (!data) return <EmptyList />;
  return (
    <View style={styles.container}>
      <TextInput style={styles.input} value={searchTerm} onChangeText={setSearchTerm} testID="searchInput" />
      <JokeList {...data} />
    </View>
  );
}

const styles = StyleSheet.create({
  input: {
    height: 40,
    margin: 12,
    borderWidth: 1,
    padding: 10,
    width: 300,
  },
  container: {
    paddingTop: 30,
    flex: 1,
    alignItems: "center",
  },
});
