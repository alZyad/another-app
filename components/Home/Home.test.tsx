import React from "react";
import { fireEvent, render, screen, userEvent, waitFor } from "@testing-library/react-native";
import Home from "../Home/Home";
import nock from "nock";
import { defaultSearch, emptySearch, fullSearch, oneResultSearch } from "./Home.mock";

describe("> Home", () => {
  beforeAll(() => {
    nock("https://v2.jokeapi.dev")
      .get("/joke/Programming?contains=&amount=10")
      .reply(200, defaultSearch)
      .persist()
      .get("/joke/Programming?contains=how&amount=10")
      .reply(200, fullSearch)
      .get("/joke/Programming?contains=xyz&amount=10")
      .reply(200, emptySearch)
      .get("/joke/Programming?contains=cat&amount=10")
      .reply(200, oneResultSearch);
  });

  test("renders correctly", () => {
    const tree = render(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("shows all results when filter is empty", async () => {
    render(<Home />);
    const searchInput = await screen.findByTestId("searchInput");
    const user = userEvent.setup();
    await user.type(searchInput, "");
    const results = await screen.findAllByTestId("jokePreview");
    expect(results).toHaveLength(10);
  });

  test("filters correctly search term 'xyz'", async () => {
    render(<Home />);
    await waitFor(() => expect(screen.getAllByTestId("jokePreview")).toHaveLength(10));
    const searchInput = await screen.findByTestId("searchInput");
    fireEvent.changeText(searchInput, "xyz");
    await waitFor(
      () => {
        expect(screen.queryAllByTestId("jokePreview")).toHaveLength(0);
      },
      { timeout: 20000 }
    );
  }, 20000);

  test("filters correctly search term 'how'", async () => {
    render(<Home />);
    const searchInput = await screen.findByTestId("searchInput");
    let results = await screen.findAllByTestId("jokePreview");
    expect(results).toHaveLength(10);

    fireEvent.changeText(searchInput, "how");

    results = await screen.findAllByTestId("jokePreview");
    expect(results).toHaveLength(10);
  });

  test("filters correctly search term 'cat'", async () => {
    render(<Home />);
    let results = await screen.findAllByTestId("jokePreview");
    expect(results).toHaveLength(10);
    const searchInput = await screen.findByTestId("searchInput");
    fireEvent.changeText(searchInput, "cat");
    await waitFor(
      () => {
        expect(screen.getAllByTestId("jokePreview")).toHaveLength(1);
      },
      { timeout: 20000 }
    );
  }, 20000);
});
