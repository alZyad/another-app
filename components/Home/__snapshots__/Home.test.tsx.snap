// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`> Home renders correctly 1`] = `
<View
  style={
    {
      "alignContent": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "textAlign": "center",
      }
    }
  >
    No jokes here, look away
  </Text>
</View>
`;
